lib/inertia_rails/renderer.rb

```diff
@@ -24,15 +14,19 @@ def initialize(component, controller, request, response, render_method, **option
 
       @controller = controller
       @configuration = controller.__send__(:inertia_configuration)
-      @component = resolve_component(component)
       @request = request
       @response = response
       @render_method = render_method
-      @props = options.fetch(:props, component.is_a?(Hash) ? component : controller.__send__(:inertia_view_assigns))
       @view_data = options.fetch(:view_data, {})
-      @deep_merge = options.fetch(:deep_merge, configuration.deep_merge_shared_data)
-      @encrypt_history = options.fetch(:encrypt_history, configuration.encrypt_history)
+      @encrypt_history = options.fetch(:encrypt_history, @configuration.encrypt_history)
       @clear_history = options.fetch(:clear_history, controller.session[:inertia_clear_history] || false)
+
+      deep_merge = options.fetch(:deep_merge, @configuration.deep_merge_shared_data)
+      passed_props = options.fetch(:props, component.is_a?(Hash) ? component : @controller.__send__(:inertia_view_assigns))
+      @props = merge_props(shared_data, passed_props, deep_merge)
```

Comment:
This line is the fix... merging props immediately so that downstream methods can reference `@props` and have the full set of props. This seems like the least surprising behavior.